### PR TITLE
add target_prefix decoding feature

### PR DESCRIPTION
**Pull request by Zenglinxiao**
In this PR, the prefix decoding feature is added, enable us to generate the prediction begin with provided lines as prefix.
To use this feature, we need to pass the prefix file as `-tgt`, and specify `-tgt_prefix` to enable this feature.

### CODE REVIEW

**Comment by francoishernandez** [comment]
Looks good, thanks a lot!

@guillaumekln would you mind having a quick look at this if you've got some time? Thanks!

**Comment by guillaumekln** [comment]
Was this inspired by an existing implementation somewhere? This seems to work but I think there is another way that is simpler and more efficient:

1. Forward the target prefix (excluding the last token) in the decoder in a teacher forcing mode and return the last decoder state. It's even possible to skip logits computation for extra performance.
2. Run beam search from the last token of the prefix and the state returned in 1. (both should be tiled the beam size). 
3. Prepend all returned hypotheses by the prefix.

**Comment by Zenglinxiao** [comment]
> Was this inspired by an existing implementation somewhere? This seems to work but I think there is another way that is simpler and more efficient:
> 
>     1. Forward the target prefix (excluding the last token) in the decoder in a teacher forcing mode and return the last decoder state. It's even possible to skip logits computation for extra performance.
> 
>     2. Run beam search from the last token of the prefix and the state returned in 1. (both should be tiled the beam size).
> 
>     3. Prepend all returned hypotheses by the prefix.

No, this is my original implementation.
Yes, I initially intend to do what you've said, but later I realize it will cause some issue, as the prefix we feed is not the same length. Therefore in a batch, there would be <eos> in the middle of padded sentence followed with <pad>. Therefore, we can not exclude last token to remove <eos> for those shorter sentence. In the target_prefix regium, I don't think it's a good idea to allow <eos> <pad> existe in the pass. As it will impact the hidden states which may make the prediction of following step weird.

**Comment by guillaumekln** [comment]
Right, batching with variable lengths makes it more complex.

In CTranslate2 I made the assumption that this feature is mostly used in an interactive scenario where batch mode is irrelevant. So the code just runs with batch_size = 1 on examples with a target prefix.

For other cases, I guess your approach is fine.

**Comment by francoishernandez** [comment]
I think it can be useful to have the ability to run with batch_size > 1 for experiments purposes.
One can always use CT2 for more efficient inference anyways.

**Comment by medicine-wave** [comment]
how can I use the ```target_prefix ```? 
https://github.com/OpenNMT/OpenNMT-py/issues/2483
